File: minecraft-web-manager/backend/src/api/controllers/auth.controller.js
import jwt from 'jsonwebtoken';
import User from '../../models/User.js';
import Session from '../../models/Session.js';
import AuditLog from '../../models/AuditLog.js';

const JWT_SECRET = process.env.JWT_SECRET || 'change-this-secret';
const JWT_EXPIRES_IN = process.env.JWT_EXPIRES_IN || '2h';
const REFRESH_TOKEN_EXPIRES_IN = process.env.REFRESH_TOKEN_EXPIRES_IN || '7d';

class AuthController {
  /**
   * POST /api/auth/login
   * Login de usuario
   */
  static async login(req, res) {
    try {
      const { username, password } = req.body;

      // Validar campos requeridos
      if (!username || !password) {
        return res.status(400).json({
          success: false,
          message: 'Username y password son requeridos'
        });
      }

      // Buscar usuario
      const user = User.findByUsername(username);
      if (!user) {
        return res.status(401).json({
          success: false,
          message: 'Credenciales inválidas'
        });
      }

      // Verificar contraseña
      const isValidPassword = User.verifyPassword(password, user.password_hash);
      if (!isValidPassword) {
        return res.status(401).json({
          success: false,
          message: 'Credenciales inválidas'
        });
      }

      // Generar tokens
      const token = jwt.sign(
        { userId: user.id, username: user.username, role: user.role },
        JWT_SECRET,
        { expiresIn: JWT_EXPIRES_IN }
      );

      const refreshToken = jwt.sign(
        { userId: user.id, type: 'refresh' },
        JWT_SECRET,
        { expiresIn: REFRESH_TOKEN_EXPIRES_IN }
      );

      // Calcular fecha de expiración
      const expiresAt = new Date();
      expiresAt.setHours(expiresAt.getHours() + 2); // 2 horas

      // Guardar sesión
      Session.create({
        userId: user.id,
        token,
        refreshToken,
        expiresAt: expiresAt.toISOString(),
        ipAddress: req.ip,
        userAgent: req.get('user-agent')
      });

      // Actualizar último login
      User.updateLastLogin(user.id);

      // Registrar en audit log
      AuditLog.create({
        userId: user.id,
        action: 'login',
        resourceType: 'auth',
        resourceId: null,
        details: { username },
        ipAddress: req.ip
      });

      // Responder
      res.json({
        success: true,
        message: 'Login exitoso',
        data: {
          token,
          refreshToken,
          user: {
            id: user.id,
            username: user.username,
            role: user.role
          }
        }
      });
    } catch (error) {
      console.error('Error en login:', error);
      res.status(500).json({
        success: false,
        message: 'Error en el servidor'
      });
    }
  }

  /**
   * POST /api/auth/logout
   * Logout de usuario
   */
  static async logout(req, res) {
    try {
      const token = req.token;
      const userId = req.user.userId;

      // Eliminar sesión
      Session.delete(token);

      // Registrar en audit log
      AuditLog.create({
        userId,
        action: 'logout',
        resourceType: 'auth',
        resourceId: null,
        details: null,
        ipAddress: req.ip
      });

      res.json({
        success: true,
        message: 'Logout exitoso'
      });
    } catch (error) {
      console.error('Error en logout:', error);
      res.status(500).json({
        success: false,
        message: 'Error en el servidor'
      });
    }
  }

  /**
   * POST /api/auth/refresh
   * Renovar token con refresh token
   */
  static async refresh(req, res) {
    try {
      const { refreshToken } = req.body;

      if (!refreshToken) {
        return res.status(400).json({
          success: false,
          message: 'Refresh token requerido'
        });
      }

      // Verificar refresh token
      let decoded;
      try {
        decoded = jwt.verify(refreshToken, JWT_SECRET);
      } catch (error) {
        return res.status(401).json({
          success: false,
          message: 'Refresh token inválido o expirado'
        });
      }

      // Buscar sesión
      const session = Session.findByRefreshToken(refreshToken);
      if (!session) {
        return res.status(401).json({
          success: false,
          message: 'Sesión no encontrada'
        });
      }

      // Buscar usuario
      const user = User.findById(decoded.userId);
      if (!user) {
        return res.status(401).json({
          success: false,
          message: 'Usuario no encontrado'
        });
      }

      // Generar nuevo token
      const newToken = jwt.sign(
        { userId: user.id, username: user.username, role: user.role },
        JWT_SECRET,
        { expiresIn: JWT_EXPIRES_IN }
      );

      // Calcular nueva fecha de expiración
      const newExpiresAt = new Date();
      newExpiresAt.setHours(newExpiresAt.getHours() + 2);

      // Actualizar sesión
      Session.updateToken(refreshToken, newToken, newExpiresAt.toISOString());

      res.json({
        success: true,
        message: 'Token renovado',
        data: {
          token: newToken
        }
      });
    } catch (error) {
      console.error('Error en refresh:', error);
      res.status(500).json({
        success: false,
        message: 'Error en el servidor'
      });
    }
  }

  /**
   * GET /api/auth/me
   * Obtener usuario actual
   */
  static async me(req, res) {
    try {
      const userId = req.user.userId;

      // Buscar usuario
      const user = User.findById(userId);
      if (!user) {
        return res.status(404).json({
          success: false,
          message: 'Usuario no encontrado'
        });
      }

      res.json({
        success: true,
        data: {
          id: user.id,
          username: user.username,
          role: user.role,
          createdAt: user.created_at,
          lastLogin: user.last_login
        }
      });
    } catch (error) {
      console.error('Error en me:', error);
      res.status(500).json({
        success: false,
        message: 'Error en el servidor'
      });
    }
  }

  /**
   * PUT /api/auth/password
   * Cambiar contraseña
   */
  static async changePassword(req, res) {
    try {
      const userId = req.user.userId;
      const { currentPassword, newPassword } = req.body;

      if (!currentPassword || !newPassword) {
        return res.status(400).json({
          success: false,
          message: 'Contraseña actual y nueva contraseña requeridas'
        });
      }

      // Validar longitud de contraseña
      if (newPassword.length < 6) {
        return res.status(400).json({
          success: false,
          message: 'La nueva contraseña debe tener al menos 6 caracteres'
        });
      }

      // Buscar usuario
      const user = User.findById(userId);
      if (!user) {
        return res.status(404).json({
          success: false,
          message: 'Usuario no encontrado'
        });
      }

      // Verificar contraseña actual
      const isValidPassword = User.verifyPassword(currentPassword, user.password_hash);
      if (!isValidPassword) {
        return res.status(401).json({
          success: false,
          message: 'Contraseña actual incorrecta'
        });
      }

      // Cambiar contraseña
      User.changePassword(userId, newPassword);

      // Registrar en audit log
      AuditLog.create({
        userId,
        action: 'change_password',
        resourceType: 'auth',
        resourceId: null,
        details: null,
        ipAddress: req.ip
      });

      res.json({
        success: true,
        message: 'Contraseña cambiada exitosamente'
      });
    } catch (error) {
      console.error('Error en changePassword:', error);
      res.status(500).json({
        success: false,
        message: 'Error en el servidor'
      });
    }
  }
}

export default AuthController;

File: minecraft-web-manager/backend-python/app/api/controllers/auth_controller.py
from datetime import datetime, timedelta
from typing import Optional
from fastapi import HTTPException, status, Response, Request
from sqlalchemy.orm import Session

from app.core.security import verify_password, create_access_token, create_refresh_token, decode_token
from app.models.user import User
from app.models.session import Session as UserSession
from app.models.audit_log import AuditLog
from app.schemas.auth import TokenResponse


class AuthController:
    @staticmethod
    async def login(
        username: str,
        password: str,
        response: Response,
        request: Request,
        db: Session
    ) -> TokenResponse:
        """Login de usuario y creación de sesión"""
        
        # Buscar usuario
        user = db.query(User).filter(User.username == username).first()
        if not user or not verify_password(password, user.password_hash):
            raise HTTPException(
                status_code=status.HTTP_401_UNAUTHORIZED,
                detail="Credenciales inválidas"
            )
        
        # Crear tokens
        access_token = create_access_token({"sub": str(user.id), "role": user.role})
        refresh_token = create_refresh_token({"sub": str(user.id)})
        
        # Crear sesión en DB
        expires_at = datetime.utcnow() + timedelta(days=7)
        session = UserSession(
            user_id=user.id,
            token=access_token,
            refresh_token=refresh_token,
            expires_at=expires_at,
            ip_address=request.client.host,
            user_agent=request.headers.get("user-agent", "")
        )
        db.add(session)
        
        # Actualizar last_login
        user.last_login = datetime.utcnow()
        
        # Audit log
        audit = AuditLog(
            user_id=user.id,
            action="login",
            resource_type="auth",
            ip_address=request.client.host,
            details={"username": username}
        )
        db.add(audit)
        
        db.commit()
        
        # Establecer cookie HttpOnly
        response.set_cookie(
            key="access_token",
            value=access_token,
            httponly=True,
            max_age=7200,  # 2 horas
            samesite="lax"
        )
        
        return TokenResponse(
            access_token=access_token,
            token_type="bearer",
            user={"id": user.id, "username": user.username, "role": user.role}
        )
    
    @staticmethod
    async def logout(user_id: int, request: Request, response: Response, db: Session):
        """Logout y eliminación de sesión"""
        
        # Obtener token de la cookie
        token = request.cookies.get("access_token")
        
        if token:
            # Eliminar sesión de DB
            db.query(UserSession).filter(UserSession.token == token).delete()
            db.commit()
        
        # Audit log
        audit = AuditLog(
            user_id=user_id,
            action="logout",
            resource_type="auth",
            ip_address=request.client.host
        )
        db.add(audit)
        db.commit()
        
        # Eliminar cookie
        response.delete_cookie("access_token")
        
        return {"message": "Logout exitoso"}
    
    @staticmethod
    async def refresh_token(
        old_token: str,
        response: Response,
        db: Session
    ) -> TokenResponse:
        """Renovar access token usando refresh token"""
        
        # Buscar sesión
        session = db.query(UserSession).filter(
            UserSession.refresh_token == old_token
        ).first()
        
        if not session or session.expires_at < datetime.utcnow():
            raise HTTPException(
                status_code=status.HTTP_401_UNAUTHORIZED,
                detail="Refresh token inválido o expirado"
            )
        
        # Obtener usuario
        user = db.query(User).filter(User.id == session.user_id).first()
        if not user:
            raise HTTPException(
                status_code=status.HTTP_401_UNAUTHORIZED,
                detail="Usuario no encontrado"
            )
        
        # Crear nuevo access token
        new_access_token = create_access_token({"sub": str(user.id), "role": user.role})
        
        # Actualizar sesión
        session.token = new_access_token
        db.commit()
        
        # Actualizar cookie
        response.set_cookie(
            key="access_token",
            value=new_access_token,
            httponly=True,
            max_age=7200,
            samesite="lax"
        )
        
        return TokenResponse(
            access_token=new_access_token,
            token_type="bearer",
            user={"id": user.id, "username": user.username, "role": user.role}
        )
    
    @staticmethod
    async def get_current_user(user_id: int, db: Session):
        """Obtener información del usuario actual"""
        user = db.query(User).filter(User.id == user_id).first()
        if not user:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Usuario no encontrado"
            )
        
        return {
            "id": user.id,
            "username": user.username,
            "role": user.role,
            "created_at": user.created_at,
            "last_login": user.last_login
        }
